arangod/RocksDBEngine/RocksDBReplicationContext.cpp

```diff
@@ -1120,49 +1114,26 @@ void RocksDBReplicationContext::use(double ttl) {
   }
 }
 
-void RocksDBReplicationContext::release() {
-  MUTEX_LOCKER(locker, _contextLock);
-  TRI_ASSERT(_users > 0);
-  TRI_ASSERT(_ttl > 0.0);
-  _expires = TRI_microtime() + _ttl;
-  --_users;
-
-  // make sure the WAL files are not deleted immediately
-  std::set<TRI_vocbase_t*> dbs;
-  for (auto& pair : _iterators) {
-    dbs.emplace(&pair.second->vocbase);
-  }
-  for (TRI_vocbase_t* vocbase : dbs) {
-    vocbase->replicationClients().track(syncerId(), replicationClientServerId(),
-                                        clientInfo(), _snapshotTick, _ttl);
-  }
-}
-
-/// extend without using the context
-void RocksDBReplicationContext::extendLifetime(double ttl) {
-  MUTEX_LOCKER(locker, _contextLock);
-  ttl = std::max(_ttl, ttl);
-  TRI_ASSERT(ttl > 0.0);
-  _expires = TRI_microtime() + ttl;
-}
-
 /// create rocksdb snapshot, must hold _contextLock
 void RocksDBReplicationContext::lazyCreateSnapshot() {
   _contextLock.assertLockedByCurrentThread();
   if (_snapshot == nullptr) {
-    _snapshot = _engine.db()->GetSnapshot();
-    TRI_ASSERT(_snapshot);
-    _snapshotTick = _snapshot->GetSequenceNumber();
+    _snapshot = std::make_shared<rocksdb::ManagedSnapshot>(_engine.db());
+
+    TRI_ASSERT(_snapshot != nullptr);
+    TRI_ASSERT(_snapshot->snapshot() != nullptr);
+    _snapshotTick = _snapshot->snapshot()->GetSequenceNumber();
   }
 }
 
 RocksDBReplicationContext::CollectionIterator::CollectionIterator(
     TRI_vocbase_t& vocbase, std::shared_ptr<LogicalCollection> const& coll,
-    bool sorted, rocksdb::Snapshot const* snapshot)
+    bool sorted, std::shared_ptr<rocksdb::ManagedSnapshot> snapshot)
```

Comment:
What is the difference between ManagedSnapshot and Snapshot?

Also it looks like it can be unique_ptr, but maybe I'm wrong

Reply:
A ManagedSnapshot is a RAII implementation from RocksDB, that automatically acquires a snapshot upon creation, and release it upon destruction.
So it is not necessary to call `CreateSnapshot()` and `ReleaseSnapshot()` on the DB anymore.

In theory we could use a unique_ptr here, but as callers can request the iterator's snapshot, I could either return a raw pointer or a shared_ptr. I decided to use the latter one, because the code is not performance-critical, and I would like avoid the use of raw pointers here (that was part of the motivation for this PR).

Reply:
Ok